internal/controller/observedobjectcollection/reconciler.go

```diff
@@ -0,0 +1,318 @@
+)
+
+const (
+	errNewKubernetesClient = "cannot create new Kubernetes client"
+	errStatusUpdate        = "cannot update status"
+	fieldOwner             = client.FieldOwner("observed-object-collection-controller")
+	membershipLabelKey     = "kubernetes.crossplane.io/owned-by-collection"
+)
+
+// Reconciler watches for ObservedObjectCollection resources
+// and creates observe-only Objects for the matched items.
+type Reconciler struct {
+	client             client.Client
+	log                logging.Logger
+	pollInterval       func() time.Duration
+	clientForProvider  func(ctx context.Context, inclusterClient client.Client, providerConfigName string) (client.Client, error)
+	observedObjectName func(collection client.Object, matchedObject client.Object) (string, error)
+}
+
+// Setup adds a controller that reconciles ObservedObjectCollection resources.
+func Setup(mgr ctrl.Manager, o controller.Options, pollJitter time.Duration) error {
+	name := managed.ControllerName(v1alpha1.ObservedObjectCollectionGroupKind)
+
+	r := &Reconciler{
+		client: mgr.GetClient(),
+		log:    o.Logger,
+		pollInterval: func() time.Duration {
+			return o.PollInterval + +time.Duration((rand.Float64()-0.5)*2*float64(pollJitter)) //nolint
+		},
+		clientForProvider:  clients.ClientForProvider,
+		observedObjectName: observedObjectName,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.ObservedObjectCollection{}).
+		WithEventFilter(predicate.Or(
+			predicate.GenerationChangedPredicate{},
+			predicate.AnnotationChangedPredicate{},
+			predicate.LabelChangedPredicate{}),
+		).
+		Complete(ratelimiter.NewReconciler(name, xperrors.WithSilentRequeueOnConflict(r), o.GlobalRateLimiter))
+}
+
+// Reconcile fetches objects specified by their GVK and label selector
+// and creates observed-only Objects for the matches.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, error error) { //nolint:gocyclo
+	log := r.log.WithValues("request", req)
+
+	defer func() {
+		if error == nil {
+			log.Info("Reconciled")
+		} else {
+			log.Info("Retry", "err", error)
+		}
+	}()
+
+	c := &v1alpha1.ObservedObjectCollection{}
+	err := r.client.Get(ctx, req.NamespacedName, c)
+
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if meta.WasDeleted(c) {
+		return ctrl.Result{}, nil
+	}
+
+	if meta.IsPaused(c) {
+		c.Status.SetConditions(xpv1.ReconcilePaused())
+		return ctrl.Result{}, errors.Wrap(r.client.Status().Update(ctx, c), errStatusUpdate)
+	}
+
+	log.Info("Reconciling")
+
+	mlv, err := membershipLabelValue(c)
+	if err != nil {
+		werr := errors.Wrapf(err, "error getting value for label %v", membershipLabelKey)
+		c.Status.SetConditions(xpv1.ReconcileError(werr))
+		_ = r.client.Status().Update(ctx, c)
+		return ctrl.Result{}, werr
+	}
```

Comment:
Should we follow the same as crossplane-runtime? https://github.com/crossplane/crossplane-runtime/blob/991de68a3f6e90e2a5ee59067697048db333c0b3/pkg/reconciler/managed/reconciler.go#L903-L917

Reply:
What is the added value? The current approach:
* Indicates the real error on condition
* tries to update the status collection resource, but if it fails is fully irrelevant for reconciliation retry, controller-runtime gets the real error back
* the real error is logged [here](https://github.com/crossplane-contrib/provider-kubernetes/pull/217/files#diff-cbd20e87d322d219672285e5d13eb7d14481d19bc89343232c569801055895b8R98-R104)

Pasting the referred code here:

```golang
if err != nil {
		log.Debug("Cannot observe external resource", "error", err)
		if kerrors.IsConflict(err) {
			return reconcile.Result{Requeue: true}, nil
		}
		record.Event(managed, event.Warning(reasonCannotObserve, err))
		managed.SetConditions(xpv1.ReconcileError(errors.Wrap(err, errReconcileObserve)))
		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
	}
```

If I understand correctly, the real error is not going to be returned back to the caller in the above approach.

Reply:
>What is the added value? 

Not much other than following upstream. 

Sounds good to me.